### PR TITLE
backup meta_state & refactor on server_state & modify shell script when running multi meta

### PR DESCRIPTION
**Pull request by shengofsun**
1. Add code in meta_server, to support the backup of meta_state in remote storage. If you add proper arguments in [apps.meta], it will do backup and restore. config-backup-tool.ini is the example config file. This resolves issue #50 
2. modify server_state, to use the transaction api when do initializing.
3. modify the shell script multi-meta.sh, which can support run multiple meta servers and kill them in turn.

### CODE REVIEW

**Comment by imzhenyu** [comment]
@shengofsun nice work. You may want to integrate the dump feature into cli by using dsn_cli_register etc. so that we can use local|remote|web cli to do the dump work, which helps debugging.

File: src/apps/replication/meta_server/server_state.cpp [hunk 1]
```diff
@@ -398,73 +547,34 @@ error_code server_state::initialize_apps()
...
-        _apps.push_back(app);
+        std::shared_ptr<TEntries> entries = _storage->new_transaction_entries(app.partition_count+1);
+        marshall_json(data, app, app.status==app_status::available);
+        entries->create_node(get_app_path(app), data);
+        for (const partition_configuration& pc: app.partitions) {
+            marshall_json(data, pc);
+            entries->create_node(get_partition_path(pc.gpid), data);
+        }
+        error_code ec;
+        task_ptr tsk = _storage->submit_transaction(
```
Comment (imzhenyu) [inline src/apps/replication/meta_server/server_state.cpp:569]: Is there upper bound about the size of the transaction? I'm worried that the partition count can be very large in real settings, e.g., tens of thousands.

File: src/apps/replication/meta_server/dump_file.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            else {
+                derror("read file failed, offset(%ld)", ftell(_file_handle));
+                return -1;
+            }
+        }
+
+        std::shared_ptr<char> ptr(new char[hdr.length], [](char* raw){ delete []raw; });
+        char* raw_mem = ptr.get();
+        len = 0;
+        while (len < hdr.length)
```
Comment (qinzuoyan) [inline src/apps/replication/meta_server/dump_file.h:96]: the 'while' may cause endless loop

File: src/apps/replication/meta_server/dump_file.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        dassert(_is_write, "call append when open file with read mode");
+        block_header hdr = {data_length, 0};
+        hdr.crc32 = dsn_crc32_compute(data, data_length, 0);
+        int len = fwrite(&hdr, sizeof(hdr), 1, _file_handle);
+        if (len < 1) {
+            derror("append file failed");
+            return -1;
+        }
+        len = 0;
+        while (len < data_length)
```
Comment (qinzuoyan) [inline src/apps/replication/meta_server/dump_file.h:62]: the 'while' may cause endless loop

File: src/apps/replication/meta_server/server_state.cpp (hunk 1 above)
Comment (shengofsun) [inline src/apps/replication/meta_server/server_state.cpp:569]: this is indeed an issue which I didn't notice

**Comment by imzhenyu** [comment]
@shengofsun you may want to fix and close issue https://github.com/imzhenyu/rDSN/issues/115 as well since you are already working on related issues:-)

**Pull request merged**